### PR TITLE
[clang-repl] Extend the C support.

### PR DESCRIPTION
**Pull request by vgvassilev**
The IdResolver chain is the main way for C to implement lookup rules. Every new partial translation unit caused clang to exit the top-most scope which in turn cleaned up the IdResolver chain. That was not an issue for C++ because its lookup is implemented on the level of declaration contexts.

This patch keeps the IdResolver chain across partial translation units maintaining proper C-style lookup infrastructure.

### CODE REVIEW

File: clang/test/Interpreter/execute.c
Comment (weliveindetail) [inline clang/test/Interpreter/execute.c:10]: Interesting. Not sure it's relevant, but the difference in IR is the missing `(...)` in front of the callee token:
Screenshot 2024-04-24 at 17 58 49

Does this mean something like "unknown actual parameters"?

It also happens with a `static` function, but here we emit the function body right away (and we know the actual args in IR):
Screenshot 2024-04-24 at 18 05 46

I guess this is not the whole story though. Printf looks like this:
```
define internal void @__stmts__0() #0 {
entry:
  %call = call i32 (ptr, ...) @printf(ptr noundef @.str, i32 noundef 2)
  ret void
}

declare i32 @printf(ptr noundef, ...) #1
```

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (weliveindetail) [inline clang/lib/Interpreter/IncrementalParser.cpp:416]: In which case wouldn't we have a frontend token here?

**Review by weliveindetail — COMMENTED** [review]
I am not an expert on the behavior of IdResolver, but this patch works for me.

File: clang/test/Interpreter/execute.c
Comment (vgvassilev) [reply clang/test/Interpreter/execute.c:10]: Not a C expert, my guess was that it had to do with CodeGen deciding to defer the inline as being a weak symbol. Needs more debugging but that's probably something for a later stage.

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:416]: If the `MostRecentTU` has a declaration which for some reason was removed from the IdResolver by exiting some lexical scope, for example? I suspect there are other language rules that may decide to make a declaration not visible to the C name lookup...

**Comment by vgvassilev** [comment]
The pre-merge windows test failure in `Clang :: Driver/amdgpu-toolchain.c` seems unrelated to this PR.

**Comment by vgvassilev** [comment]
@AaronBallman, can you take a look at that patch, hopefully to move forward as seems the other reviewers are busy.

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (weliveindetail) [reply clang/lib/Interpreter/IncrementalParser.cpp:416]: Alright, so in this case `IdResolver` has no entry for `ND`? Can we assert that? Might also be a good a pointer for the uninformed reader.

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (AaronBallman) [inline clang/lib/Interpreter/IncrementalParser.cpp:415]:
```suggestion
    const auto *ND = dyn_cast<NamedDecl>(D);
    if (!ND)
      continue;
    // Check if we need to clean up the IdResolver chain.
```

File: clang/lib/Sema/SemaDecl.cpp
Comment (AaronBallman) [inline clang/lib/Sema/SemaDecl.cpp:2286]: so in incremental processing mode, we never remove declarations from the id resolver, so they always stay available? e.g.,
```
{
  int i = 12;
}
printf("%d", i); // Fine?
```

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:415]: The `isa` + `cast` saves a few instructions in the case where `D` is not a `NamedDecl`. I'd prefer to stay the way it is although we are not expecting a lot of declarations that are not `NamedDecl`s.

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:416]: To clarify, you want an assert for the case where we try to remove the `ND` from the `IdResolved` but it was never registered? If so, `RemoveDecl` asserts (I was hurt couple of times there ;))

File: clang/lib/Sema/SemaDecl.cpp
Comment (vgvassilev) [reply clang/lib/Sema/SemaDecl.cpp:2286]: No, this change makes sure that we preserve the state of the `IdResolver` across partial translation units (PTU). That is, when we see another incremental output, we still take into account the declarations that came from previous PTUs. Eg:

```
[repl] int i = 12;
[repl] printf("%d\n", i); // without this patch we report we do not know what `i` is...
```

To your explicit question this patch produces:

```
bin/clang-repl -Xcc -xc
clang-repl> { int i = 12; }
clang-repl> printf("%d", i); // Fine?
In file included from <<< inputs >>>:1:
input_line_2:1:1: error: call to undeclared library function 'printf' with type 'int (const char *, ...)'; ISO C99 and later do not support implicit function declarations [-Wimplicit-function-declaration]
    1 | printf("%d", i); // Fine?
      | ^
input_line_2:1:1: note: include the header <stdio.h> or explicitly provide a declaration for 'printf'
input_line_2:1:14: error: reference to local variable 'i' declared in enclosing context
    1 | printf("%d", i); // Fine?
      |              ^
input_line_1:1:7: note: 'i' declared here
    1 | { int i = 12; }
      |       ^
error: Parsing failed.
clang-repl> %quit
```

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (AaronBallman) [reply clang/lib/Interpreter/IncrementalParser.cpp:415]: isa + cast is a code smell because that is literally doing what dyn_cast does; do you have measurements that suggest there is a measurable performance difference?

File: clang/lib/Sema/SemaDecl.cpp
Comment (AaronBallman) [reply clang/lib/Sema/SemaDecl.cpp:2286]: Ah thank you for the explanation -- I think that warrants a comment in the code to explain what's going on, because it's not obvious from context.

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:415]: Hm. Ok I did not realize that code is smelly :)

We are getting a bit theoretical here but for the sake of the argument:

`isa<>` calls [CastInfo<To, const From>::isPossible](https://llvm.org/doxygen/structllvm_1_1CastIsPossible.html#a44e6fa462c45ccec1dcb2e52901e055d)(Val);

`cast<>` calls [CastInfo<To, const From>::doCast](https://llvm.org/doxygen/structllvm_1_1CastInfo.html#a143f6f757917e1245aed3882209b466f)(Val);

`dyn_cast<>` is both `isa + cast`.

My point was that in the case of something that's not a `NamedDecl` we will avoid the check `isPossible` for that case. It's probably peanuts: https://godbolt.org/z/xeh7jh5ax

Anyway, if you feel strongly about the other approach I can do it...

Comment (AaronBallman) [reply clang/lib/Interpreter/IncrementalParser.cpp:415]: If we have performance numbers that suggest this is a hot path and we need the split, then we can keep it. But in general, we [document](https://llvm.org/docs/ProgrammersManual.html#the-isa-cast-and-dyn-cast-templates) that particular pattern:

> "Note that you should not use an isa<> test followed by a cast<>, for that use the dyn_cast<> operator."

Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:415]: I don't. Let me make these changes so we can move forward here.

Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:415]: OTOH, that loop is easier to be spelled like:
```
for() {
  if (auto *ND = dyn_cast<NamedDecl>(D))
    // do some work
}
```

That violates the loop early exit policy :)

File: clang/lib/Sema/SemaDecl.cpp
Comment (vgvassilev) [reply clang/lib/Sema/SemaDecl.cpp:2287]: Added a comment.

**Comment by vgvassilev** [comment]
The bots fail on windows due to `TEST 'Clang :: CoverageMapping/mcdc-system-headers.cpp' FAILED` which does not seem related to this PR.

**Review by AaronBallman — APPROVED** [review]
LGTM with a possible improvement in the code (take it or leave it, your call).

File: clang/lib/Interpreter/IncrementalParser.cpp
Comment (vgvassilev) [reply clang/lib/Interpreter/IncrementalParser.cpp:416]: It always good to take rather than leave ;) I haven't got used to this syntax and I'd leave it :(

**Comment by vgvassilev** [comment]
The failures on windows are due to "out of heap" error. Let's land this to test on the "real" bots.

**Comment by mysterymath** [comment]
We've started seeing crashes in Fuchsia's Mac tooclhain's LLDB test suite:
https://luci-milo.appspot.com/ui/p/fuchsia/builders/toolchain.ci/clang-mac-x64/b8747305520511321169/overview

This is the only PR in the blamelist that looks plausible, but there isn't a clear connection to me (or a clear reason why it would be Mac-only). Does anything jump out, or am I barking up the wrong tree?

Stack trace follows:
```
Assertion failed: (D->getLexicalDeclContext() == this && "Decl inserted into wrong lexical context"), function addHiddenDecl, file DeclBase.cpp, line 1692.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	/Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSZone.h:32:1 <Spelling=/Volumes/Work/s/w/ir/cache/macos_sdk/XCode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX13.3.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:382:22>: at annotation token
 #0 0x0000000111461238 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x8ab238)
 #1 0x000000011145f179 llvm::sys::RunSignalHandlers() (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x8a9179)
 #2 0x0000000111461900 SignalHandler(int) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x8ab900)
 #3 0x00007ff807c915ed (/usr/lib/system/libsystem_platform.dylib+0x7ff8004255ed)
 #4 0x0000700014de4c40 
 #5 0x00007ff807b8ab45 (/usr/lib/system/libsystem_c.dylib+0x7ff80031eb45)
 #6 0x00007ff807b89e5e (/usr/lib/system/libsystem_c.dylib+0x7ff80031de5e)
 #7 0x0000000114a4ae54 clang::DeclContext::addHiddenDecl(clang::Decl*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x3e94e54)
 #8 0x0000000114a4aeae clang::DeclContext::addDecl(clang::Decl*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x3e94eae)
 #9 0x0000000113edeb81 clang::Sema::ActOnStartTopLevelStmtDecl(clang::Scope*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x3328b81)
#10 0x0000000113a31ccf clang::Parser::ParseTopLevelStmtDecl() (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2e7bccf)
#11 0x0000000113a6d231 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2eb7231)
#12 0x0000000113a45202 clang::Parser::ParseLinkage(clang::ParsingDeclSpec&, clang::DeclaratorContext) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2e8f202)
#13 0x0000000113a6f037 clang::Parser::ParseDeclOrFunctionDefInternal(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec&, clang::AccessSpecifier) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2eb9037)
#14 0x0000000113a6e8a5 clang::Parser::ParseDeclarationOrFunctionDefinition(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*, clang::AccessSpecifier) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2eb88a5)
#15 0x0000000113a6cfd4 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2eb6fd4)
#16 0x0000000113a6ad37 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2eb4d37)
#17 0x0000000113a6a501 clang::Parser::ParseFirstTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2eb4501)
#18 0x0000000113956acb clang::ParseAST(clang::Sema&, bool, bool) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2da0acb)
#19 0x0000000113721a53 clang::FrontendAction::Execute() (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2b6ba53)
#20 0x000000011369068d clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2ada68d)
#21 0x000000011369cd24 void llvm::function_ref<void ()>::callback_fn<compileModuleImpl(clang::CompilerInstance&, clang::SourceLocation, llvm::StringRef, clang::FrontendInputFile, llvm::StringRef, llvm::StringRef, llvm::function_ref<void (clang::CompilerInstance&)>, llvm::function_ref<void (clang::CompilerInstance&)>)::$_1>(long) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x2ae6d24)
#22 0x000000011145c0fe llvm::CrashRecoveryContext::RunSafely(llvm::function_ref<void ()>) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x8a60fe)
#23 0x000000011145c220 RunSafelyOnThread_Dispatch(void*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x8a6220)
#24 0x000000011145c23f void* llvm::thread::ThreadProxy<std::__2::tuple<void (*)(void*), (anonymous namespace)::RunSafelyOnThreadInfo*>>(void*) (/Volumes/Work/s/w/ir/x/w/llvm_build/lib/liblldb.19.0.0git.dylib+0x8a623f)
#25 0x00007ff807c641d3 (/usr/lib/system/libsystem_pthread.dylib+0x7ff8003f81d3)
#26 0x00007ff807c5fbd3 (/usr/lib/system/libsystem_pthread.dylib+0x7ff8003f3bd3)
```

**Comment by jasonmolenda** [comment]
I reverted in 
```
commit dfdf1c5fe45a82b9c578306f3d7627fd251d63f8
Author: Jason Molenda <jmolenda@apple.com>
Date:   Tue May 21 18:00:11 2024 -0700

    Revert "[clang-repl] Extend the C support. (#89804)"
    
    This reverts commit 253c28fa829cee0104c2fc59ed1a958980b5138c.
```

to unblock the CI bots until Vassil has a chance to look at it.

**Comment by vgvassilev** [comment]
> I reverted in
> 
> ```
> commit dfdf1c5fe45a82b9c578306f3d7627fd251d63f8
> Author: Jason Molenda <jmolenda@apple.com>
> Date:   Tue May 21 18:00:11 2024 -0700
> 
>     Revert "[clang-repl] Extend the C support. (#89804)"
>     
>     This reverts commit 253c28fa829cee0104c2fc59ed1a958980b5138c.
> ```
> 
> to unblock the CI bots until Vassil has a chance to look at it.

Thank you! I suspect that this is due to the changes in SemaDecl.cpp (https://github.com/llvm/llvm-project/pull/89804/files#diff-edac6256ac508912a16d0165b2f8cf37123dc2f40a147dca49a34c33f1db13ddR2288)

I am trying to build lldb but if you have a hint how to reproduce it outside lldb would be great!

**Comment by vgvassilev** [comment]
@jasonmolenda, I am stuck.  I could not find how the bot configures llvm and lldb. I built lldb and ran `make check-lldb` but did not fail in the same way. Can you provide a recipe to reproduce the failure?

**Comment by mysterymath** [comment]
> @jasonmolenda, I am stuck. I could not find how the bot configures llvm and lldb. I built lldb and ran `make check-lldb` but did not fail in the same way. Can you provide a recipe to reproduce the failure?

For some reason it only fails on our Mac LLDB builders, and not our Linux ones. Our LLDB configuration is also quite tortured I'm afraid; it's difficult to reproduce locally. I'm not sure whether this issue would be seen on a clean Mac build; it seems like it might be some kind of pointer Heisenbug or something. I can definitely test a candidate fix, but that's about it.

**Comment by mysterymath** [comment]
I was able to verify that the revert fixed the tests at least.

**Comment by vgvassilev** [comment]
> > @jasonmolenda, I am stuck. I could not find how the bot configures llvm and lldb. I built lldb and ran `make check-lldb` but did not fail in the same way. Can you provide a recipe to reproduce the failure?
> 
> For some reason it only fails on our Mac LLDB builders, and not our Linux ones. Our LLDB configuration is also quite tortured I'm afraid; it's difficult to reproduce locally. I'm not sure whether this issue would be seen on a clean Mac build; it seems like it might be some kind of pointer Heisenbug or something. I can definitely test a candidate fix, but that's about it.

I tried on a mac machine. Can I get access to that bot then?

**Comment by vgvassilev** [comment]
> I was able to verify that the revert fixed the tests at least.

I am pretty sure that the failure ha to do with the incremental processing if-stmt in SemaDecl. I can provide a fix that does not trigger this for lldb but that would be a workaround. I suspect that change uncovers some inconsistency in either clang or lldb which I think we should fix.

**Comment by mysterymath** [comment]
I did have a means to reproduce this locally, but it's been a long time since I've used it. I should be able to recreate the setup. I'll get back to you with the details when I can; this whole afternoon is booked through pretty solid with meetings for me I'm afraid.

**Comment by vgvassilev** [comment]
Ok, maybe I can you test this patch for a temporary fix until we figure out what's going wrong.
```diff
diff --git a/clang/lib/Sema/SemaDecl.cpp b/clang/lib/Sema/SemaDecl.cpp
index 7f6921ea22be..d771a060f305 100644
--- a/clang/lib/Sema/SemaDecl.cpp
+++ b/clang/lib/Sema/SemaDecl.cpp
@@ -2285,7 +2285,7 @@ void Sema::ActOnPopScope(SourceLocation Loc, Scope *S) {
     // Partial translation units that are created in incremental processing must
     // not clean up the IdResolver because PTUs should take into account the
     // declarations that came from previous PTUs.
-    if (!PP.isIncrementalProcessingEnabled())
+    if (!PP.isIncrementalProcessingEnabled() && !getLangOpts().IncrementalExtensions)
       IdResolver.RemoveDecl(D);
 
     // Warn on it if we are shadowing a declaration.
```

**Comment by mysterymath** [comment]
This still exhibits the same issue: 

https://luci-milo.appspot.com/raw/build/logs.chromium.org/fuchsia/led/dthorn_google.com/0d77d2b0963e18ad11cb83db0680c2382ed7f8df95cf733f42fcdbd1dda8a7ec/+/build.proto

https://logs.chromium.org/logs/fuchsia/led/dthorn_google.com/0d77d2b0963e18ad11cb83db0680c2382ed7f8df95cf733f42fcdbd1dda8a7ec/+/u/lldb/test/stdout

I do think it should be possible to download a "source tarball" that allows easily reproducing the output of our bots on similar host configurations, but I don't think we're quite there yet. It's an AI on our end, for sure. I asked around what people were doing, and it involved basically replicating the bot build by hand using a ton of tribal knowledge and manual editing; it's not something that I could export any easier than it would be to creating a better way to do this.

**Comment by jasonmolenda** [comment]
fwiw if you have access to a mac, I don't think it's hard to build lldb+clang?  I have swig, cmake, and ninja installed on my mac via homebrew.

```
lldb/scripts/macos-setup-codesign.sh

mkdir build
cd build
 cmake ../llvm -G Ninja -DLLDB_ENABLE_SWIFT_SUPPORT=0 -DPython3_EXECUTABLE=/usr/bin/python3 -DCMAKE_BUILD_TYPE=Debug -DLLVM_ENABLE_ASSERTIONS=1 -DLLVM_ENABLE_PROJECTS='llvm;lldb;clang' '-DLLVM_ENABLE_RUNTIMES=libcxx;libcxxabi;libunwind' -DLLDB_ENABLE_PYTHON=1 -DLLDB_ENABLE_LUA=0

ninja lldb
bin/lldb-dotest -p TestImportBuiltinFileID.py
```

**Comment by junaire** [comment]
Just found out this keeps popping up from my notification, and I would like give my 2 cents, maybe can provide some insights:

Looking at the backtrace of the failing tests, I noticed we ran into `clang::Parser::ParseTopLevelStmtDecl()` which makes me very confused. Do we actually enable the incremental extension in lldb? I grep it in lldb codebase and the only occurrence is here https://github.com/llvm/llvm-project/blob/b91b8fea8c58e9b414e291df677b12ca44197784/lldb/source/Plugins/ExpressionParser/Clang/ClangModulesDeclVendor.cpp#L662

The other thing I noticed is that the failure looks like have something to do with the objc, can we also check the current language we're compiling when skipping removing the name for the lexical scope?

**Comment by vgvassilev** [comment]
> fwiw if you have access to a mac, I don't think it's hard to build lldb+clang? I have swig, cmake, and ninja installed on my mac via homebrew.
> 
> ```
> lldb/scripts/macos-setup-codesign.sh
> 
> mkdir build
> cd build
>  cmake ../llvm -G Ninja -DLLDB_ENABLE_SWIFT_SUPPORT=0 -DPython3_EXECUTABLE=/usr/bin/python3 -DCMAKE_BUILD_TYPE=Debug -DLLVM_ENABLE_ASSERTIONS=1 -DLLVM_ENABLE_PROJECTS='llvm;lldb;clang' '-DLLVM_ENABLE_RUNTIMES=libcxx;libcxxabi;libunwind' -DLLDB_ENABLE_PYTHON=1 -DLLDB_ENABLE_LUA=0
> 
> ninja lldb
> bin/lldb-dotest -p TestImportBuiltinFileID.py
> ```

Does not seem to be reproducing it for me:

```
builds/llvm-project$ bin/lldb-dotest -p TestImportBuiltinFileID.py
/Applications/Xcode.app/Contents/Developer/usr/bin/python3 /Users/vvassilev/workspace/sources/llvm-project/lldb/test/API/dotest.py --arch x86_64 --build-dir /Users/vvassilev/workspace/builds/llvm-project/lldb-test-build.noindex --executable /Users/vvassilev/workspace/builds/llvm-project/./bin/lldb --compiler /Users/vvassilev/workspace/builds/llvm-project/./bin/clang --dsymutil /Users/vvassilev/workspace/builds/llvm-project/./bin/dsymutil --lldb-libs-dir /Users/vvassilev/workspace/builds/llvm-project/./lib --llvm-tools-dir /Users/vvassilev/workspace/builds/llvm-project/./bin --libcxx-include-dir /Users/vvassilev/workspace/builds/llvm-project/include/c++/v1 --libcxx-library-dir /Users/vvassilev/workspace/builds/llvm-project/lib --lldb-obj-root /Users/vvassilev/workspace/builds/llvm-project/tools/lldb -p TestImportBuiltinFileID.py
lldb version 19.0.0git (git@github.com:llvm/llvm-project.git revision 253c28fa829cee0104c2fc59ed1a958980b5138c)
  clang revision 253c28fa829cee0104c2fc59ed1a958980b5138c
  llvm revision 253c28fa829cee0104c2fc59ed1a958980b5138c
Skipping the following test categories: ['libstdcxx', 'dwo', 'llgs', 'fork']
PASS: LLDB (/Users/vvassilev/workspace/builds/llvm-project/bin/clang-x86_64) :: test_import_builtin_fileid_gmodules (TestImportBuiltinFileID.TestImportBuiltinFileID)
----------------------------------------------------------------------
Ran 1 test in 52.311s

OK
```

**Comment by vgvassilev** [comment]
> Just found out this keeps popping up from my notification, and I would like give my 2 cents, maybe can provide some insights:
> 
> Looking at the backtrace of the failing tests, I noticed we ran into `clang::Parser::ParseTopLevelStmtDecl()` which makes me very confused. Do we actually enable the incremental extension in lldb? I grep it in lldb codebase and the only occurrence is here
> 
> https://github.com/llvm/llvm-project/blob/b91b8fea8c58e9b414e291df677b12ca44197784/lldb/source/Plugins/ExpressionParser/Clang/ClangModulesDeclVendor.cpp#L662

Hm, ok. I remember some discussions revolving around this but got the impression lldb turns that off. I think having it on is the right thing to do but that explains why my workaround did not work.

> 
> The other thing I noticed is that the failure looks like have something to do with the objc, can we also check the current language we're compiling when skipping removing the name for the lexical scope?

I am hoping to cover objc, too. That's why I am really interested in figuring out what is going on. Otherwise if we limit the feature to c/c++ maybe we will get away with the bot. I still want to see why it crashes though...

**Comment by jasonmolenda** [comment]
I set aside my current work, updated to current main sources, cherry-picked 253c28fa829cee0104c2fc59ed1a958980b5138c, built it on my macOS system, did `bin/lldb-dotest -p TestImportBuiltinFileID.py` and it crashes like before.

I'm not sure what you're asking help with, I don't know much of anything about how the clang setup in lldb is done - I happened to be watching the CI bots and saw they all started failing and reverted the change.  

You couldn't reproduce this behavior on a macOS system?  it is an arm64 macOS system, I didn't try repo'ing on an Intel box.  

Earlier on this PR you mused about making this change in SemaDecl.cpp

```
-    if (!PP.isIncrementalProcessingEnabled())
+    if (!PP.isIncrementalProcessingEnabled() && !getLangOpts().IncrementalExtensions)
```

the test case still crashes for me with that change.

**Comment by vgvassilev** [comment]
@jasonmolenda, I struggled quite a bit in reproducing the test failures and everything but I have a fix. Thanks a lot for your time and I hope the fix works!

**Comment by hokein** [comment]
@vgvassilev 

The reland d999ce0302f06d250f6d496b56a5a5f2dc331e61 makes the clang reject the valid code now:

```
$ cat /tmp/t33.cpp
#include <pthread.h>
#include <algorithm>

int main() {

}
                                                                                                                                      
$  ./bin/clang -Xclang -fsyntax-only -Xclang=-fincremental-extensions -stdlib=libc++ /tmp/t33.cpp                                      <<<
In file included from /tmp/t33.cpp:2:
In file included from /usr/include/c++/v1/algorithm:1712:
In file included from /usr/include/c++/v1/memory:877:
In file included from /usr/include/c++/v1/iterator:684:
In file included from /usr/include/c++/v1/__iterator/common_iterator.h:22:
/usr/include/c++/v1/variant:1024:43: error: use of undeclared identifier '__arg'; did you mean '_pthread_cleanup_buffer::__arg'?
 1024 |       } __impl{this, _VSTD::forward<_Arg>(__arg)};
      |                                           ^
/usr/include/pthread.h:162:9: note: '_pthread_cleanup_buffer::__arg' declared here
  162 |   void *__arg;                            /* Its argument.  */
      |         ^
In file included from /tmp/t33.cpp:2:
In file included from /usr/include/c++/v1/algorithm:1861:
In file included from /usr/include/c++/v1/__algorithm/ranges_stable_partition.h:15:
/usr/include/c++/v1/__algorithm/stable_partition.h:246:18: error: use of undeclared label '__second_half_done'
  246 |             goto __second_half_done;
      |                  ^
2 errors generated.
```

Can you take a look?

**Comment by vgvassilev** [comment]
@hokein, ah, that's annoying. Can you provide the entire proprocessed file that does not work? I'd like to bisect and debug.

**Comment by hokein** [comment]
Thanks for the prompt response. I think limiting it to C-only will fix the issue (note that there is no `C` in `LangOpts`, you may want to use `!getLangOpts().CPlusPlus` to exclude C++).

**Comment by vgvassilev** [comment]
> Thanks for the prompt response. I think limiting it to C-only will fix the issue (note that there is no `C` in `LangOpts`, you may want to use `!getLangOpts().CPlusPlus` to exclude C++).

Makes sense. If that works for you, we can check this in. I want to somehow record this breakage in the form of a test for our future selves when decide to revisit this workaround-looking code.

Out of curiosity, in what context you use `-fincremental-extensions`?

**Comment by hokein** [comment]
> Out of curiosity, in what context you use -fincremental-extensions?

The code snippet I provided is extracted from our internal test. We have an internal clang-tool (with the `incremental-extensions` on) to generate headers

**Comment by vgvassilev** [comment]
> > Out of curiosity, in what context you use -fincremental-extensions?
> 
> The code snippet I provided is extracted from our internal test. We have an internal clang-tool (with the `incremental-extensions` on) to generate headers

I am a bit overwhelmed right now, are you willing to open a PR with that change so we can merge it -- otherwise I could look later this week :(

**Comment by hokein** [comment]
> I want to somehow record this breakage in the form of a test for our future selves when decide to revisit this workaround-looking code.

This is the processed file  https://gist.github.com/hokein/e4a5881329c3956494afa2de7d350476.

> I am a bit overwhelmed right now, are you willing to open a PR with that change so we can merge it -- otherwise I could look later this week :(

Sure, I will prepare one.

**Pull request merged**
